### DISCUSSION TITLE
Fetch Protected API with protected access token in Pimcore

### DISCUSSION THREAD

**Discussion by Gunnu131** [pimcore/pimcore#15705] [Q&A / Support]
How i can make consume Protected API in pimcore through Data Importer. There is no input tab in pimcore to enter Access Token Key. Is there any other possible ways to do it?

**Comment by kingjia90** [comment]
Do you mean this 
![image](https://github.com/pimcore/pimcore/assets/6014195/14a42652-6e60-4618-a726-01b7ad440211)
https://pimcore.com/docs/platform/Data_Importer/Configuration/Data_Sources#configuration-options-3
?

**Reply by Gunnu131** [reply]
Thanks for the response. but I didnt mean by this; My data will come through HTTP source . I think The push data source is a mechanism that allows data to be sent to a remote location via a POST request.  But I want to recieve the data in pimcore

**Reply by kingjia90** [reply]
AFAIU The push is a receiving endpoint, for `push` is meant that is pushed from somewhere into pimcore, therefore is a receiving data mechanism.

**Reply by Gunnu131** [reply]
Ohh Okayy So in push where should I've to define the API Url. For Example :- https:// 10.1.45.23/test/testing.php ....Where to fill this input so pimcore get knows from where to fetch data?

**Reply by kingjia90** [reply]
as in the screenshot and docs, the endpoint field is what you might looking for eg. `https://******/pimcore-datahub-import/car-import/push`

**Reply by Gunnu131** [reply]
The endpoint field is ineditable. This field is showing where the data will come in pimcore but we can't define the endpoint from where to fetch the data

**Reply by kingjia90** [reply]
Sounds like that in https:// 10.1.45.23/test/testing.php there should be a curl POST with the access token or so to that data hub importer endpoint.  Not sure if i got it right

**Reply by Gunnu131** [reply]
<?php
// Define the valid bearer token
$valid_bearer_token = 'Token Id';

// Check if the Authorization header is set and the bearer token is valid
$headers = getallheaders();
if (!isset($headers['Authorization']) || $headers['Authorization'] !== 'Bearer ' . $valid_bearer_token) {
    // Send an error message and exit if the token is invalid or missing
    http_response_code(401); // Unauthorized
    echo json_encode(array("message" => "Invalid or missing bearer token."));
    exit;
}

// Database connection settings
$servername = "localhost";
$username = "username";
$password = "password";
$dbname = "test";

// Create a connection to the database
$conn = new mysqli($servername, $username, $password, $dbname);

// Check the connection
if ($conn->connect_error) {
    die("Connection failed: " . $conn->connect_error);
}

// Define the API response format
$response = array();

// Fetch data from the database
$sql = "SELECT * FROM test";
$result = $conn->query($sql);

// Check if any data was found
if ($result->num_rows > 0) {
    // Loop through the result and store data in the response
    while ($row = $result->fetch_assoc()) {
        $response[] = $row;
    }

    // Return the data as JSON
    echo json_encode($response);
} else {
    // If no data found, return an error message
    echo json_encode(array("message" => "No data found."));
}

// Close the database connection
$conn->close();
?>

Thats my API which is fetching data from xampp mysql server and i dont know how to push this protected api in pimcore bcz the endpoint in push is ineditable

**Reply by kingjia90** [reply]
Please generate a token with the thunder icon (screenshot) and pass it to the curl
https://reqbin.com/req/c-hlt4gkzd/curl-bearer-token-authorization-header-example#:~:text=To%20send%20a%20Bearer%20Token,control%20access%20to%20protected%20resources.

Unless the data importer is clearly bugged/broken, i am afraid i won't be able to help beyond that. 
The attached code looks fine (typical php vanilla db fetch) even though may missing some parts from copy/paste.
